### ISSUE TITLE
[Engine/ScriptNode] Local Cable Hijacks The Network Parameter via ID

### ISSUE THREAD

**Issue by aaronventure** [christophhart/HISE#647]
Local Cable is able to change the actual network parameter i.e. it can send data upstream if the name of the cable matches that of the parameter
```
HiseSnippet 1338.3oc2X0sSibCE1SBlPxRqJUUauqJWxVgPYRxF.0K3m.gMpafncnnsWQMy3PrxL1ilwCPZUuuu.8UnWWgp5ssW1mf9bvavV6wSxL4OfjTP6tAIzbrOG6Oe99NGOIM8XlXeelGPK2Iccw.skgFco71UaiHTP88AZeJrAxmi8xqFZuttHeerEPSK8gxAzxt.H7ysauGxFQMwwCA.mxHl3WSbH73Qaty2RrsqgrvmPbR3c4cpaxnUY1r.AdRCK.bQlcPWfOBIcKED7Jjeaf1WCK7xJX7FkZYYgJUxpRkJas4Vk2RG0pPYz45U1pbwxaTbyMZAzV7.KBm4YvQbruXQ2iY00nM6JpZCNk3SN2FKMzAFhcVMLnZahsUydIGe.PaglwopzpT0W.aPrH8GONk8YgSjONhjIMsT2EjzmBHok.RKnfzJPCSOhKOdFIddFrNUvfsPBtIITT9Bz9aXUlvAJecGTGbMOgQ+.VsRgBqkW7uW7MsBnlbBilmQOhwwGSW8E49obYy8y4xO7TsZM14jaiGy1F6M1okxAu6JvUoANmi8VK+kH6.beGEG+AyoK9vxolpScBGYz5TB+XWbjcMlskLWIedTF.Dk1DO8c02GwQRRIZLgetXONQBGs8wWJJCTTTV39X+NblqnPXD9SnbXVA1H9fxIYgVzDh7w.bnjnn9Dd2jEhSgFqvcpwdnPbEXSB2r83wXpwfQQl5w.iQUleB7fVsvl7X.t.r1am0xvBScYXtnxPwdJ27kiLyW6s5is.7WfwhcWOrKxCeBqoMp6p9HGWa7aDXbs7mayL6XP9Q7nUHtJHrmziUMainTrs+rTHs3+SMm.ugEvIzKZf3djqAZviBbLD80MwUiPmXLsTxZHkcAosTWXfoVgFuS7IZRcosVzj58lLQ42QX9ULuNgzQzy.sLpbueXR9rVWqC101lcUUliKIRkJ3fvwZxr651lQIlxgTdzCo65vBDaRDbEWAcBhXKk0FA9hFEVGSMDNGdumV5Swd9gKbFXg0E+Ij4GwrDSkoFxTjX61DwaKqWj8dDhSr25l8EoCf0j5D0EfCLsbU6cUoJwIPsiLaHunTRYMQdhvDswkcfRokNQGIknUZIJISo1559mJm0DYCNUJJjAEtKR2gpCwhRWeNTnxP1mYhDJgyBn9BCALG378UPOE+u9X8czy1DVzjmxrK9pa902M7I8dNWYguVtt0s5cndFzf4fqScCD5l39Umf7t.y8iZfSwlJ4QH4IBV7VPtNnXb0O2JmKCTs1fDYbIR5YAzVpAgFs+wM.aftdnwtcaCN1UVgmzOiN3qTI2jdJOb813QV58wsPA17QFe2.NyQTYa0mbi3zLPkJ7dDoixa8B7tkiym3SKAPWRse5SOR6G4iITSk.pCIYF.vOODvhFwqOfOih5gVjoA6Sg59KGa0W4GhLO868x7D.Lyn.71+Z9AXNXCALHt1ciw3xucq+3e9y+c6wiwa29I.VP3tVVO7rVx9ASRPLeM4mzp9z2k+CsV0Ig6fF2Koo+nPZ5CQZ276GN2j1xwjVwoj0x7dBqc+zQwGE5n3Pzwg27ayMcrRLcXnPyGqbRoGENozPbxOby7Wh74CyIevVnLYDp9Eqh6eOlKG2dTD+867vQ7HK2jtjN4qKM42dZBe2kw+xS6wXcbPgeU0Y6GM5o3KV6fL8XmE8JpxC9RgiHjqzPAeVXCocdcvkCSrNDKxYllCtTiDXwYMvRyZfkm0.e4rFXkYMvMl0.279CT9CSG8pERQrn1u4AghPMsCnR8aXaMv+AlhAny
```

![local cable messes with parameter](https://github.com/user-attachments/assets/12e2b51b-f936-4265-982e-77c603e4e8a2)

It doesn't happen if the name/id is different, which points to the id being the issue.
![local cable messes with parameter 2](https://github.com/user-attachments/assets/fb5f38ed-59d6-4bb2-9361-c3a6abbfa5b4)

**Comment by christoph-hart** [comment]
haha, interesting, if you set the multiply parameter to something slightly less than 1.0, then the parameter knob decays slowly towards 1. Definitely the bug of the year 2025 so far as I wouldn't know what might cause it from the top of my head.

**Comment by christoph-hart** [comment]
OK, so it seems that it's not related to the parameters having the same name - I can reproduce it with a different network parameter ID.

You have a recursive loop in the local cable - it sets the value of the pma node, which multiplies it and sets the same cable value to a slightly lower number which then gets send back to the original cable. If you remove the parameter connection to the cable and drag the first node, you'll end up with the same procedure but without the UI glitching (which comes from the parameter knob being updated asynchronously).

Now the fact that the parameter is working "upstream" has nothing to do with the parameter IDs being identical (which would be weird because they are organized in two completely separate systems), but an optimization of scriptnode which detects if a network parameter is a direct 1:1 connection to a single parameter with the same range and in this case just mirrors its value. You can quickly remove the "upstream" functionality by connecting any other target to the network parameter - as soon as it's not a single 1:1 connection, it will manage its own value and thus stop responding as local target.

The most simple scenario to replicate this would be this here:

![Image](https://github.com/user-attachments/assets/0a558c29-91f8-4eb4-8e93-8515b134c16d)

```
HiseSnippet 1217.3ocyWEraaaDDcojWKK0jfDf1a8.yMmBCCQWEaGzCN1xVsBMRVHz0vnWLVubk0BQtKA4R6nVz68iom5WP+F5sh9SDzefzYHojnrjcLDRMpMfs1Ylc4adyalkpWjlKhi0QDqZmLJTPrdD0cjxLn4.lTQZeHw5IzNrXiHxNyzAiBYwwBOhkU4uEMXUcER5OueuCX9LEWL0DgbpVxEuQFHMSs160euz2uEySbhLnPzMdcatV0T6qS.7TlVmDx3CYWJ5xvvJQIeGKd.w5qndNMb5uKms8tNM1hy2Z2c15U6vEr98Ea67xcZra+F8Y75aQrV8HOoQG4ZXFQLwZkCzdibGnuVk8.NUFKuvWfKbHtvSNybKsuGlhnURyAReudiIpXBbJ8lRakynsOm1Q5ImXeJ88zTG1S2QQBzpzrvq7LvyoH7pW.dK.RVEfzJYP5YTWdjLzL0ChmOi1VAUSfdDy.krXIV+AsoFBPY1LfMTzJBVLYCquc85aXC+4EeS+DE2H0Jaspq1HNVs9Kp8y0pV6WpYeSW86uPe3iIR66KhVnaTZDcWabcURvEhnMruh4mHlDHj9yxozamSKVx4YYcg.0p1Jo43Pg51DBjbpB9zOz9PlggEhbaPbghHiDgf0ghqf1frxRU5gh3gFcHzHLWMCTKZuDelYVID1nk6.3fYpaXwQEKMiJ1H9ISWceg3yn8jF9fEiwRK.i.S8eAFy6FeL8HXT.2LEfqPac1CWqWs7VO3YhO7Gkuzt0YNKro6WoSE3gQhPVj3DcOe1n0iYAg9h2BXbC6K707gtxeRLeWQXFDN.iXc9.lRI7iWllmUu2rhycWTdqNwHUW1gYhjuC5B6lD3By04hl4nCrYUB6gxVWGWi5BWgxKcwGfexc5fqsxc5L1Yg1utBy05ngoki7OSrpfbeY5Px9995qapCBk4hSf5Ss0S6OJbfVI4norHFCv8CzIvYmiR3lmSXReTM6lDCyD7NV4BAmdcmU4SEQwoGbEZ8MgeA0cWsG3pRKFG3yQ8XlAXaBNlAzjhnM4SzlHDKpJxttCshmw36CyXG.iAXJi2Fh0kdrHHZX9LNlojU4BicxTl3JnuqTlNrc7onWNymbJV4wMk9TvvKmA4UwPoTlmGYFvWiF.+eSz97nMK96.u8v6UTPOocfv1nsMCD1S.OZ.utwVZrigQi1Qh3PsxCDPnKPUy7s4LP4Yih6KEwOGRoh4Jo.QfI93UDq05HU445zgOcXu6F1d+dtFQH1cULN2ghqyHghQhDTEZ1AL2QennOKw2Lm88SLZf.yHMRAp9wzz767z7aVJ+KnQYsQaVLj4Y+YOhhUgpq15269gapb9H5jpz2fGWauwYWEZqD0vQ.uMY.+IrnKEl+Wv7+ye14hmd1es28g4KUj4eRQZy4Af5qA2cdk.5.d9CTInxCTIXq09su7G896asDbGHDOu0nY4B4SObumPDdmIq7AT.yFm+hb4KymlCEiaTtmjGnqwjB4.sdX.K81nk68BeHt6Lfwizmyydc.LeWK0BnKUo55pvW6CVa6Pt5lEj.3KzbNmO6QM2F2ZY23WurarwxtwWtrab6kci6rrab2O9FwuuY9HOT6BMP8NJUDZYcjBksY2.8uXA0Z0A
```

TLDR, I would classify this as a user error rather than a bug in the system - if you create recursions by dragging cables in a circle you'll suppose to end up with a non-functional network and with HISE you have to be grateful that it didn't delete your hard drive but just glitch around on the UI.

Best I can add do is to a few UX changes to prevent this from irritating the user:

1. Add an error message if it detects a cycle. 
2. Detect whether the parameter connection is to a global or local cable node and then prevent it from making this 1:1 connection, this would then remove the "upstream" changing of the parameter. In all other cases / nodes the connected parameter target cannot be changed from the outside so it can't create an "upstream" value change to the network parameter.

**Comment by aaronventure** [comment]
But how can anything from the network affect the main network parameters? I thought these were only meant to be controlled from the script or directly by the user.

**Comment by christoph-hart** [comment]
> But how can anything from the network affect the main network parameters? I thought these were only meant to be controlled from the script or directly by the user.

This is a side effect of an optimization of scriptnode: If there is a direct 1:1 connection without scaling or range conversion from a network parameter to a node parameter, then the network parameter directly links to the parameter as target (so it bypasses one redirection). So basically in this case there is no network parameter and the knob of the network parameter (or if you change it externally) directly controls the node's parameter. If you change that value, the main parameter also changed, because it's the same thing. Usually this never occurs because you cannot change a connected parameter, but with local / global cables you can so this creates the false impression of the local cable changing something it shouldn't.

Anyways, I've changed this optimization to be skipped when connecting it to a value parameter of the global or local cables as well as macro parameters of nested containers (this caused also a few UI glitches).

Let's consider that solved :)
